File: 7. Reverse Integer/solution.py
# Runtime: 32 ms, faster than 98.76% of Python3 online submissions for Reverse Integer.
# Memory Usage: 13.2 MB, less than 56.92% of Python3 online submissions for Reverse 


class Solution:
    def reverse(self, x: int) -> int:
        isNeg = 1 if x > 0 else -1
        x *= isNeg
        s = str(x)[::-1]
        res = int(s) * isNeg
        if abs(res) >= (2 ** 31 - 1): return 0
        return res

File: 7. Reverse Integer/solution.cpp
/*
    Runtime: 4 ms, faster than 95.57% of C++ online submissions for Reverse Integer.
    Memory Usage: 8.3 MB, less than 66.26% of C++ online submissions for Reverse Integer.
*/

class Solution {
public:
    int reverse(int x) {
        if (x == INT_MIN) return 0;
        int isNeg = (x < 0) ? -1 : 1;
        x *= isNeg;
        
        int res = 0;
        while(x != 0){
            if (res > INT_MAX/10) return 0;
            if (res == INT_MAX/10 && x%10 > 7) return 0;
            res *= 10;
            res += x % 10;
            x /= 10;
        }
        res *= isNeg;
        return res;
    }
};
